File: life-js/life.mjs
import sleepSynchronously from 'sleep-synchronously';

const width = 40;
const height = 20;
const iterations = 100;
const BENCHMARK = false;

let startState = [
    [false, false, false, false, false],
    [false, false, true, false, false],
    [false, false, false, true, false],
    [false, true, true, true, false],
    [false, false, false, false, false],
];

function updateState(oldState, newState) {
    for (let y = 0; y < oldState.length; y = y + 1) {
        for (let x = 0; x < oldState[y].length; x = x + 1) {
            let count = countNeighbors(oldState, x, y);
            newState[y][x] = count === 3 || (oldState[y][x] && count === 2);
        }
    }
}

function printState(state) {
    if (BENCHMARK)
        return;
    clearScreen();
    for (let row of state) {
        let buf = '';
        for (let cell of row) {
            if (cell) {
                buf = buf + '*';
            } else {
                buf = buf + '.';
            }
        }
        console.log(buf);
    }
}

function delay(ms) {
    if (BENCHMARK)
        return;
    sleepSynchronously(ms);
}

function readState(state, x, y) {
    if (y < 0) {
        y = y + state.length
    }
    if (x < 0) {
        x = x + state[0].length
    }
    if (y >= state.length) {
        y = 0
    }
    if (x >= state[0].length) {
        x = 0
    }
    return state[y][x];
}

function countNeighbors(state, x, y) {
    let count = 0;
    for (let row = y - 1; row <= y + 1; row = row + 1) {
        for (let col = x - 1; col <= x + 1; col = col + 1) {
            if (row === y && col === x)
                continue;
            if (readState(state, col, row))
                count = count + 1;

        }
    }
    return count;
}

function createMatrix(w, h) {
    let result = Array(h);
    for (let i = 0; i < h; ++i) {
        result[i] = Array(w).fill(false);
    }
    return result;
}

function clearScreen() {
    process.stdout.write('\u001b[2J\u001b[0;0H');
}

function runSimulation(oldState, steps) {
    let newState = createMatrix(oldState[0].length, oldState.length);
    for (let i = 0; i < steps; i = i + 1) {
        updateState(oldState, newState);

        printState(newState);
        delay(50);

        let tmp = oldState;
        oldState = newState;
        newState = tmp;
    }
    return oldState;
}

function main() {
    // Resize startState.
    if (startState.length < height) {
        for (let i = startState.length; i < height; ++i)
            startState[i] = Array();
        for (let row of startState) {
            for (let i = row.length; i < width; ++i)
                row[i] = false;
        }
    }

    printState(startState);
    runSimulation(startState, iterations);
}

main();

File: life-python/life.py
import os
import numpy as np
import time

height = 40
width = 20
# Change BEMCHMARK to True and iterations to 50000 test the performance
iterations = 100
BENCHMARK = False

startState = np.zeros((width, height), dtype=bool)


def updateState(oldState, newState):
    for row in range(0, len(oldState)):
        for col in range(0, len(oldState[row])):
            count = countNeighbors(oldState, row, col)
            newState[row][col] = count == 3 or (
                oldState[row][col] and count == 2)
    return newState


def countNeighbors(state, x, y):
    count = 0
    for row in range(-1, 2):
        for col in range(-1, 2):
            if row == 0 and col == 0:
                continue
            if readState(state, x + row, y + col):
                count = count + 1
    return count


def readState(state, x, y):
    if x < 0:
        x = x + len(state)
    if y < 0:
        y = y + len(state[0])
    if x >= len(state):
        x = 0
    if y >= len(state[0]):
        y = 0

    return state[x][y]


def printState(state):
    if BENCHMARK:
        return
    os.system('clear')
    time.sleep(0.05)
    for row in state:
        buf = ""
        for cell in row:
            if cell:
                buf = buf + "*"
            else:
                buf = buf + "."
        print(buf)


def runSimulation(oldState, steps):
    newState = np.zeros((width, height), dtype=bool)
    for i in range(0, steps):
        newState = updateState(oldState, newState)
        printState(newState)
        tmp = oldState
        oldState = newState
        newState = tmp
    return newState


def main():
    # Splicing the array to input automata seeds
    startState[1][2] = True
    startState[2][3] = True
    startState[3][1:4] = True
    # startState[2][7:10] = True
    printState(startState)
    runSimulation(startState, iterations)


main()
